Microsoft import/export and cloud extensions

jimmarino opened the pull request:
I bundled everything up into one commit so the merge goes smoother.

seehamrun commented on extensions/auth/portability-auth-microsoft/src/main/java/org/dataportabilityproject/auth/microsoft/MicrosoftAuthServiceProvider.java:
```diff
@@ -0,0 +1,23 @@
+/**
+ *
+ */
+public class MicrosoftAuthServiceProvider implements AuthServiceProvider {
```
Could you explain a bit how/where the MicrosoftAuthServiceProvider will register itself with the AuthServiceProviderRegistry?

jimmarino replied:
Really good question! We need to discuss this as the second part of modularity: how are runtimes assembled with extensions. My understanding is that currently "uber" jars are created but I think this may become difficult to maintain if people want to create and deploy extensions that are not part of the repository. 

There are probably a number of ways we can approach this. For example, using Guice and a bootstrapper, JPMS ServiceLoaders, classpath ServiceLoaders, something else, or a combination of the above. Most of these approaches will involve each extension having some type of bootstrap assembler class/provider that is discovered and invoked. For testing, fixtures would do this work.

To accommodate this, I left the assembly and registration of these extensions for later until we address those issues.  So basically, assembly will be done by some external mechanism TBD. 

Let me know if you have some thoughts on what might work here.

seehamrun replied:
Yes currently we have an Uber jar and the ServiceProviderRegistry which registers and loads serviceProviders based on the api settings file. https://github.com/google/data-portability/blob/1944be15d1753da6dc63681f5de7216707ea0b41/portability-core/src/main/java/org/dataportabilityproject/ServiceProviderRegistry.java#L55

For a first version, we can probably do something similar and discuss the dynamic loading requirements later after everything is ported over.

jimmarino replied:
Sounds good!

rtannenbaum commented on extensions/cloud/portability-cloud-microsoft/src/main/java/org/dataportabilityproject/cloud/microsoft/cosmos/CassandraCluster.java:
```diff
@@ -0,0 +1,89 @@
+    private Cluster cluster;
+    private String host = "127.0.0.1";
+    private int port = 10350;
+    private String username = "localhost";
```
If we run the APIs/workers on Kubernetes on Azure, like we do for Google Cloud Platform, we expose credentials to access GCP services in a Kubernetes secret: https://github.com/google/data-portability/blob/master/config/k8s/api-deployment.yaml#L53

jimmarino replied:
On Azure they have a Vault service and I imagine people may also want to use other things such as Kubernetes. I haven't had a chance to look at it yet but was Brian working on an interface to plug different secrets stores in? I could write and extension for the Microsoft service.

jimmarino commented on extensions/auth/portability-auth-microsoft/src/main/java/org/dataportabilityproject/auth/microsoft/ParamStringBuilder.java:
```diff
@@ -0,0 +1,52 @@
+package org.dataportabilityproject.auth.microsoft;
+
```
I agree we should avoid reinventing the wheel here but my concern would be bringing in a library for one class. This may go back to the issue about modularity. In my mental model, the Microsoft extensions may be running in a process by themselves (e.g. transfer worker, customized auth system, etc) so they may not have libraries used by other extensions available to them.

Is the Apache library something we need to have loaded, in which case we should definitely use it, or is it something that is optional?

holachuy replied:
Agreed, if we don't use the library for anything else, we may just want to roll our own solution if it's a few cases. 

Since it's very general, should this exist in a utilities package and we can swap it out if need be later? 

Also, Guava doesn't seem to be getting this functionality soon. Another option is OKHttp https://github.com/google/guava/issues/1005

jimmarino replied:
I'd be happy to move this up to an SPI package used by transfer, auth, and cloud extensions. 

I had a question about OKHttp. I used that in the MS extension out of habit since I've used the library in the past with success. I quickly tried the Google HTTP library used by other extensions but ran into a few issues which were probably the result of me not spending the time figuring out how to use it. In any event, maybe we want to switch over to the Google library? I figured it would be good to first get things working since extensions used by different modules will become easier to detect once the project is moved over to the new structure.

seehamrun replied:
Agreed on rolling our own solution and moving this to a shared utils directory later. A suggestion for later use to make this take a Map<String, List<String>>. That way it mimics the apache library and would avoid the state keeping transitions.

jimmarino replied:
Quick Q, where would that method go? Here's how I'm using the builder:

https://gist.github.com/jimmarino/ec864ae4570d170ac7d83ed1d2b194ae

jimmarino commented on extensions/auth/portability-auth-harness-microsoft/src/main/java/org/dataportabilityproject/auth/microsoft/harness/AuthTestDriver.java:
```diff
@@ -0,0 +1,89 @@
+        authRetrievalUrl = callbackBase + "/code";
+    }
+
+    public static void main(String... args) throws Exception {
```
I agree. That's just for me to "manually" test until I can come up with proper auth stubs (testing OAuth is a pain when your internet connection is reliant on a cheap modem that doesn't support port forwarding :(). 

I can remove it now, wait until later to remove it, or document that it will be removed. Let me know what you think works best.

rtannenbaum replied:
Gotcha, fine with me to either leave a TODO to move/remove it, or move it to a test directory

rtannenbaum replied:
If you want to make this an integration test, looks like the recommended approach in gradle is to make a separate "source dir" for integration tests rather than lumping it in with unit tests https://guides.gradle.org/migrating-from-maven/?_ga=2.198959343.1349228063.1517424826-1039443295.1516912197#integration_tests (just came across this as I was reading about gradle)

jimmarino replied:
Thanks for the pointer. I like the ideas of separate phases and dependencies for integration tests.

My original plan was just to remove this method since it is really only designed as a temporary way to test against a live MS instance to get the code in a state where it works.

Moving forward, I'd like to get proper integration tests in place for this extension:

1. A set of fast running unit tests that can be executed as part of any build without opening network connections (if possible, there will be some exceptions for localhost) or other resources. I should be able to stub out the OKHttp client to do this and have a connection that is driven by a local mock. I was going to take a similar approach with the Cassandra driver for the store (which already uses the SCassandra stub).

2. At some point we should do different levels of end-to-end integration testing. One would be testing individual extensions against test accounts, which the Gradle approach in the link will come in handy for. Another type of integration test would be to launch a complete system (workers and storage) and drive transfer. The easiest way to do this would probably be to build a test runner or Gradle plugin that bootstraps the data transfer system with provided config. If we created a Gradle plugin, other people could use it for internal testing as well.

holachuy replied:
We can add these suggestions to the existing test issue or have multiple ones for each kind of test.

jimmarino commented on extensions/cloud/portability-cloud-microsoft/src/main/java/org/dataportabilityproject/cloud/microsoft/cosmos/CassandraCluster.java:
```diff
@@ -0,0 +1,89 @@
+ */
+public class CassandraCluster {
+    private Cluster cluster;
+    private String host = "127.0.0.1";
```
Yeah, this needs to go as soon as I understand how configuration is obtained. Is there an API that provides access to the YAML properties as opposed to getResources() from the classpath?

This may be part of the further modularity discussions. For example,  a way for extensions to obtain references to SPI services and keyed properties during bootstrap? Also, are there already conventions around namespacing configuration keys so extensions don't collide?

rtannenbaum replied:
Yeah, currently we copy the yaml settings into our jar at $SRC_DIR/src/main/resources (see https://github.com/google/data-portability/blob/master/bin/build_and_run_jar.sh#L40) and expose to java code via CommonSettings (https://github.com/google/data-portability/blob/c9f95e82b44eaee627562ddb543462272fef6f3a/portability-core/src/main/java/org/dataportabilityproject/PortabilityCoreModule.java#L63)

Does that model make sense with the new design?

rtannenbaum replied:
Forgot to mention we are somewhat modular in how we do this now by selectively including different yaml files based on which binary. e.g. for the API, we include api.yaml as well as common.yaml. For the worker, we include only common.yaml (we don't have any worker-only settings as of yet).

If we want to keep using the build script in the new design, we can continue to selectively load in settings based on service provider, cloud provider and so on.

jimmarino replied:
This is probably part of the ongoing modularity discussion. 

I think we should be able to handle runtime assembly (getting the bits into a file system layout and then possibly into a container image) from Gradle.  

My preference would be to have a bootstrap API for each subsystem that takes CommonSettings as a parameter. Runtime launchers could then decide how to source configuration, e.g. YAML, environment properties, build script params, a combination, etc. There would be a "provided" launcher that used YAML but custom ones could be written, That would give users flexibility in how they want to mange deployment and configuration.

holachuy replied:
Feel free to add this as an issue

jimmarino commented on extensions/cloud/portability-cloud-microsoft/src/main/java/org/dataportabilityproject/cloud/microsoft/cosmos/CosmosStore.java:
```diff
@@ -0,0 +1,154 @@
+    static final String DATA_INSERT = "INSERT INTO  " + DATA_TABLE + " (data_id, data_model) VALUES (?,?)";
+    static final String DATA_QUERY = "SELECT * FROM " + DATA_TABLE + " WHERE data_id = ?";
+    static final String DATA_DELETE = "DELETE FROM " + DATA_TABLE + " WHERE data_id = ?";
+    static final String DATA_UPDATE = "UPDATE " + DATA_TABLE + "SET data_model = ? WHERE data_id = ?";
```
Thanks!. I'll fix that.

jimmarino commented on extensions/cloud/portability-cloud-microsoft/src/main/java/org/dataportabilityproject/cloud/microsoft/cosmos/Initializer.java:
```diff
@@ -0,0 +1,56 @@
+    /**
+     * Returns a new {@link CosmosStore} instance configured for Azure.
+     */
+    public CosmosStore get(ObjectMapper mapper) {
```
The get() is a pretty bad name. This was another part I wanted to change around based on what we do around module bootstrapping. We should talk about whether we want to require extensions to use Guice (as opposed to the core runtime). I don't yet have an opinion either way. 

In the meantime, how about if I separate the initializer out into separate classes?

jimmarino replied:
Also, to answer the original question, it's not used yet. 

My idea is to get the extension services in early and then iteratively integrate them into the runtime. There are still a number of missing pieces before we can get the extensions bootstrapped.

holachuy replied:
+1 I'm thinking we can definitely iterate with the implementations of the new architecture and flush out the runtimes over the course of a few PR's, perhaps in the next several days?  I think each of us is handling a migration of  some part of the system so we'll have many more of these discussions to sort out in the process.

The original binaries are still there for testing and fallback for demos.

rtannenbaum replied:
Cool, SGTM to get this impl in as is for now and iterate later on how we will integrate these into the architecture

jimmarino replied:
OK sounds good

jimmarino commented on extensions/auth/portability-auth-microsoft/src/main/java/org/dataportabilityproject/auth/microsoft/MicrosoftAuthDataGenerator.java:
```diff
@@ -0,0 +1,130 @@
+ * The second step takes the authorization code and posts it against the AD API to obtain a token for querying the Graph API.
+ */
+public class MicrosoftAuthDataGenerator implements AuthDataGenerator {
+    public static final String AUTHORIZATION_URL = "https://login.microsoftonline.com/common/oauth2/v2.0/authorize";
```
Yep

seehamrun commented on extensions/auth/portability-auth-microsoft/build.gradle:
```diff
@@ -0,0 +1,25 @@
+/*
```
Thanks for the explanation! This makes sense!

holachuy commented on extensions/auth/portability-auth-harness-microsoft/src/main/java/org/dataportabilityproject/auth/microsoft/harness/AuthTestDriver.java:
```diff
@@ -0,0 +1,89 @@
+import static java.lang.System.getProperty;
+
+/**
+ *
```
Fill in comment

holachuy commented on extensions/auth/portability-auth-harness-microsoft/src/main/java/org/dataportabilityproject/auth/microsoft/harness/AuthTestDriver.java:
```diff
@@ -0,0 +1,89 @@
+package org.dataportabilityproject.auth.microsoft.harness;
```
License header

holachuy commented on extensions/auth/portability-auth-harness-microsoft/src/main/java/org/dataportabilityproject/auth/microsoft/harness/AuthTestDriver.java:
```diff
@@ -0,0 +1,89 @@
+        // get the token
+        TokenAuthData tokenData = dataGenerator.generateAuthData(callbackBase, authCode, "1", configuration.getInitialAuthData(), null);
+
+        System.out.println("TOKEN: " + tokenData.getToken());
```
Should we bring in the logging framework here and use it.

jimmarino replied:
Yeah these are just temporary debug statements. 

One thing I would like to talk about is externalizing logging. This would involve defining an interface and having the code performing the process bootstrap (e.g. a launcher) provide an implementation. For example, the interface could be something like this:

https://gist.github.com/jimmarino/2652d5adbb3216eaed302637d0f15870

Classes that need to log data would declare the interface as a dependency and have it injected or provided by the instantiating class. 

This would allow people to easily create their own logging implementations (for example, one that uses JDK logging or a cloud logging service, or something else) without having to deal with trying to bind things to SLF4J or writing a JDK logging handler.

holachuy commented on extensions/auth/portability-auth-harness-microsoft/src/main/java/org/dataportabilityproject/auth/microsoft/harness/AuthTestDriver.java:
```diff
@@ -0,0 +1,89 @@
+            }
+            String authCode = new String(authBody.bytes());
+
+            System.out.println("AUTH: " + authResponse.code() + ":" + authCode);
```
likewise, same loggin gquestion.

holachuy commented on extensions/auth/portability-auth-harness-microsoft/src/main/java/org/dataportabilityproject/auth/microsoft/harness/TestHelper.java:
```diff
@@ -0,0 +1,66 @@
+import java.security.cert.X509Certificate;
+
+/**
+ *
```
Fill in comment

holachuy commented on extensions/auth/portability-auth-microsoft/src/main/java/org/dataportabilityproject/auth/microsoft/MicrosoftAuthDataGenerator.java:
```diff
@@ -0,0 +1,130 @@
+    public static final String AUTHORIZATION_URL = "https://login.microsoftonline.com/common/oauth2/v2.0/authorize";
+    public static final String TOKEN_URL = "https://login.microsoftonline.com/common/oauth2/v2.0/token";
+
+    private String redirectPath;
```
Let's mark these final if they are only set in the constructor

holachuy commented on extensions/auth/portability-auth-microsoft/src/main/java/org/dataportabilityproject/auth/microsoft/MicrosoftAuthServiceProvider.java:
```diff
@@ -0,0 +1,23 @@
+    private MicrosoftAuthDataGenerator contactsGenerator;
+
+    public MicrosoftAuthServiceProvider() {
+        //contactsGenerator = new MicrosoftAuthDataGenerator(r)
```
This should probably have a TODO or uncommented.

holachuy commented on extensions/auth/portability-auth-microsoft/src/main/java/org/dataportabilityproject/auth/microsoft/MicrosoftAuthServiceProvider.java:
```diff
@@ -0,0 +1,23 @@
+    }
+
+    public AuthDataGenerator getAuthDataGenerator(String transferDataType) {
+        return null;
```
Same here, TODO?

holachuy commented on extensions/auth/portability-auth-microsoft/src/main/java/org/dataportabilityproject/auth/microsoft/ParamStringBuilder.java:
```diff
@@ -0,0 +1,52 @@
+package org.dataportabilityproject.auth.microsoft;
```
License header

holachuy commented on extensions/auth/portability-auth-microsoft/src/main/java/org/dataportabilityproject/auth/microsoft/MicrosoftAuthServiceProvider.java:
```diff
@@ -0,0 +1,23 @@
+package org.dataportabilityproject.auth.microsoft;
```
license header

holachuy commented on extensions/auth/portability-auth-microsoft/src/main/java/org/dataportabilityproject/auth/microsoft/ParamStringBuilder.java:
```diff
@@ -0,0 +1,52 @@
+        START, START_PARAM, PARAM_VALUE, END_PARAM
+    }
+
+    private StringBuilder builder = new StringBuilder();
```
final

holachuy commented on extensions/auth/portability-auth-microsoft/src/test/java/org/dataportabilityproject/auth/microsoft/ParamStringBuilderTest.java:
```diff
@@ -0,0 +1,40 @@
+import org.junit.Test;
+
+/**
+ *
```
Let's fill in comment

holachuy commented on extensions/auth/portability-auth-microsoft/src/test/java/org/dataportabilityproject/auth/microsoft/ParamStringBuilderTest.java:
```diff
@@ -0,0 +1,40 @@
+package org.dataportabilityproject.auth.microsoft;
```
license header

holachuy commented on extensions/cloud/portability-cloud-microsoft/src/main/java/org/dataportabilityproject/cloud/microsoft/cosmos/CassandraCluster.java:
```diff
@@ -0,0 +1,89 @@
+ * Configures a client connection to a Cassandra cluster.
+ */
+public class CassandraCluster {
+    private Cluster cluster;
```
final

holachuy commented on extensions/cloud/portability-cloud-microsoft/src/main/java/org/dataportabilityproject/cloud/microsoft/cosmos/CassandraCluster.java:
```diff
@@ -0,0 +1,89 @@
+ */
+public class CassandraCluster {
+    private Cluster cluster;
+    private String host = "127.0.0.1";
```
static

holachuy commented on extensions/cloud/portability-cloud-microsoft/src/main/java/org/dataportabilityproject/cloud/microsoft/cosmos/CassandraCluster.java:
```diff
@@ -0,0 +1,89 @@
+package org.dataportabilityproject.cloud.microsoft.cosmos;
```
license header

holachuy commented on extensions/cloud/portability-cloud-microsoft/src/main/java/org/dataportabilityproject/cloud/microsoft/cosmos/CosmosStore.java:
```diff
@@ -0,0 +1,154 @@
+package org.dataportabilityproject.cloud.microsoft.cosmos;
```
license header

holachuy commented on extensions/cloud/portability-cloud-microsoft/src/main/java/org/dataportabilityproject/cloud/microsoft/cosmos/CosmosStore.java:
```diff
@@ -0,0 +1,154 @@
+    static final String DATA_DELETE = "DELETE FROM " + DATA_TABLE + " WHERE data_id = ?";
+    static final String DATA_UPDATE = "UPDATE " + DATA_TABLE + "SET data_model = ? WHERE data_id = ?";
+
+    private Session session;
```
final

holachuy commented on extensions/cloud/portability-cloud-microsoft/src/main/java/org/dataportabilityproject/cloud/microsoft/cosmos/Initializer.java:
```diff
@@ -0,0 +1,56 @@
+package org.dataportabilityproject.cloud.microsoft.cosmos;
```
license header

holachuy commented on extensions/cloud/portability-cloud-microsoft/src/main/java/org/dataportabilityproject/cloud/microsoft/cosmos/Initializer.java:
```diff
@@ -0,0 +1,56 @@
+        String dataTableQuery = "CREATE TABLE IF NOT EXISTS " + DATA_TABLE + " (data_id uuid  PRIMARY KEY, data_model text)";
+        session.execute(dataTableQuery);
+    }
+
```
extra lines

holachuy commented on extensions/cloud/portability-cloud-microsoft/src/main/java/org/dataportabilityproject/cloud/microsoft/cosmos/MicrosoftCloudConstants.java:
```diff
@@ -0,0 +1,13 @@
+package org.dataportabilityproject.cloud.microsoft.cosmos;
```
license header

holachuy commented on extensions/cloud/portability-cloud-microsoft/src/main/java/org/dataportabilityproject/cloud/microsoft/cosmos/MicrosoftCloudConstants.java:
```diff
@@ -0,0 +1,13 @@
+ * Constants used by the Microsoft cloud extension.
+ */
+public interface MicrosoftCloudConstants {
+
```
should these members be static final?

jimmarino replied:
I think they should be OK as is since interface members are by default static and final

holachuy commented on extensions/cloud/portability-cloud-microsoft/src/main/java/org/dataportabilityproject/cloud/microsoft/cosmos/MicrosoftStorageException.java:
```diff
@@ -0,0 +1,12 @@
+package org.dataportabilityproject.cloud.microsoft.cosmos;
```
license header

holachuy commented on extensions/cloud/portability-cloud-microsoft/src/test/java/org/dataportabilityproject/cloud/microsoft/cosmos/CosmosStoreTest.java:
```diff
@@ -0,0 +1,97 @@
+package org.dataportabilityproject.cloud.microsoft.cosmos;
```
license header

holachuy commented on extensions/cloud/portability-cloud-microsoft/src/test/java/org/dataportabilityproject/cloud/microsoft/cosmos/CosmosStoreTest.java:
```diff
@@ -0,0 +1,97 @@
+import static org.scassandra.matchers.Matchers.preparedStatementRecorded;
+
+/**
+ *
```
comment

holachuy commented on extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/contacts/MicrosoftContactsExporter.java:
```diff
@@ -0,0 +1,103 @@
+package org.dataportabilityproject.transfer.microsoft.contacts;
```
license header

holachuy commented on extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/contacts/MicrosoftContactsExporter.java:
```diff
@@ -0,0 +1,103 @@
+    private static final String CONTACTS_URL = "https://graph.microsoft.com/v1.0/me/contacts";
+    private static final String ODATA_NEXT = "@odata.nextLink";
+
+    private OkHttpClient client;
```
final members

holachuy commented on extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/contacts/MicrosoftContactsImporter.java:
```diff
@@ -0,0 +1,100 @@
+package org.dataportabilityproject.transfer.microsoft.contacts;
```
license header

holachuy commented on extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/contacts/MicrosoftContactsImporter.java:
```diff
@@ -0,0 +1,100 @@
+    private static final String BATCH_URL = "https://graph.microsoft.com/beta/$batch";
+    private static final String CONTACTS_URL = "me/contacts"; //must be relative for batch operations
+
+    private OkHttpClient client;
```
final members

holachuy commented on extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/contacts/MicrosoftContactsImporter.java:
```diff
@@ -0,0 +1,100 @@
+        }
+    }
+
+    private Map<String, Object> createRequest(int i, LinkedHashMap contact) {
```
static?

jimmarino replied:
What would be static here?

holachuy commented on extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/provider/MicrosoftTransferServiceProvider.java:
```diff
@@ -0,0 +1,47 @@
+public class MicrosoftTransferServiceProvider implements TransferServiceProvider {
+    private static final String CONTACTS = "contacts"; // TODO we should standardize the naming scheme if it isn't already
+
+    private Map<String, Exporter<?, ?>> exporterCache = new HashMap<>();
```
final members

holachuy commented on extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/transformer/TransformResult.java:
```diff
@@ -0,0 +1,28 @@
+package org.dataportabilityproject.transfer.microsoft.transformer;
```
license header

holachuy commented on extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/transformer/TransformResult.java:
```diff
@@ -0,0 +1,28 @@
+ * The result of a transformation operation.
+ */
+public class TransformResult<T> {
+    private List<String> problems;
```
final members

holachuy commented on extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/transformer/TransformerContext.java:
```diff
@@ -0,0 +1,33 @@
+package org.dataportabilityproject.transfer.microsoft.transformer;
```
license header

holachuy commented on extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/transformer/TransformerContext.java:
```diff
@@ -0,0 +1,33 @@
+/**
+ * Provides facilities for performing recursive transformations and collated error reporting.
+ */
+public interface TransformerContext {
```
I like this for wrapping batch operations

holachuy commented on extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/transformer/TransformerService.java:
```diff
@@ -0,0 +1,16 @@
+package org.dataportabilityproject.transfer.microsoft.transformer;
```
license header

holachuy commented on extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/transformer/TransformerServiceImpl.java:
```diff
@@ -0,0 +1,99 @@
+    }
+
+    private class TransformKey {
+        private Class<?> from;
```
final members

holachuy commented on extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/transformer/TransformerServiceImpl.java:
```diff
@@ -0,0 +1,99 @@
+package org.dataportabilityproject.transfer.microsoft.transformer;
```
license header

holachuy commented on extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/transformer/TransformerServiceImpl.java:
```diff
@@ -0,0 +1,99 @@
+import java.util.function.BiFunction;
+
+/**
+ *
```
fill in comment

holachuy commented on extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/transformer/TransformerServiceImpl.java:
```diff
@@ -0,0 +1,99 @@
+ *
+ */
+public class TransformerServiceImpl implements TransformerService {
+    Map<TransformKey, BiFunction<?, ?, ?>> cache = new HashMap<>();
```
private final

holachuy commented on extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/transformer/TransformerService.java:
```diff
@@ -0,0 +1,16 @@
+     * @param resultType the type to transform to
+     * @param input the input instance
+     */
+    <T> TransformResult<T> transform(Class<T> resultType, Object input);
```
Do we need Class<T> if we have generic T in the template?

jimmarino replied:
I think we do since the implementation needs to know what type is requested:

https://gist.github.com/jimmarino/3024ee7f1ab28f08c0d502b75820f9b6

holachuy commented on extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/transformer/common/TransformerHelper.java:
```diff
@@ -0,0 +1,40 @@
+package org.dataportabilityproject.transfer.microsoft.transformer.common;
```
license header

holachuy commented on extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/transformer/contacts/ToGraphAddressTransformer.java:
```diff
@@ -0,0 +1,30 @@
+package org.dataportabilityproject.transfer.microsoft.transformer.contacts;
```
license header

holachuy commented on extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/transformer/contacts/ToGraphContactTransformer.java:
```diff
@@ -0,0 +1,128 @@
+package org.dataportabilityproject.transfer.microsoft.transformer.contacts;
```
license header

holachuy commented on portability-types-transfer/src/main/java/org/dataportabilityproject/types/transfer/models/contacts/ContactsModelWrapper.java:
```diff
@@ -0,0 +1,21 @@
+package org.dataportabilityproject.types.transfer.models.contacts;
```
LicenseHeader

holachuy commented on portability-types-transfer/src/main/java/org/dataportabilityproject/types/transfer/models/contacts/ContactsModelWrapper.java:
```diff
@@ -0,0 +1,21 @@
+/**
+ * A collection of contacts as serialized vCards.
+ */
+public class ContactsModelWrapper extends DataModel {
```
I think in willard's PR, we changed the *Wrapper classes to *ContainerResource

jimmarino replied:
OK, I was going to make all of the suggested changes and then rebase my commits against the upstream.

jimmarino replied:
OK I'll rebase my commits once I merge the upstream in

holachuy commented on portability-types-transfer/src/main/java/org/dataportabilityproject/types/transfer/models/contacts/ContactsModelWrapper.java:
```diff
@@ -0,0 +1,21 @@
+ * A collection of contacts as serialized vCards.
+ */
+public class ContactsModelWrapper extends DataModel {
+    private String vCards;
```
final

holachuy commented on portability-types-transfer/src/main/java/org/dataportabilityproject/types/transfer/auth/TokenAuthData.java:
```diff
@@ -0,0 +1,22 @@
+ */
+@JsonTypeName("org.dataportability:TokenAuthData")
+public class TokenAuthData extends AuthData {
+    private String token;
```
final

holachuy commented on portability-types-transfer/src/main/java/org/dataportabilityproject/types/transfer/auth/TokenAuthData.java:
```diff
@@ -0,0 +1,22 @@
+package org.dataportabilityproject.types.transfer.auth;
```
license header

holachuy commented on extensions/data-transfer/portability-data-transfer-microsoft/src/test/java/org/dataportabilityproject/transfer/microsoft/integration/LocalImportTestRunner.java:
```diff
@@ -0,0 +1,66 @@
+package org.dataportabilityproject.transfer.microsoft.integration;
```
license header

holachuy commented on extensions/data-transfer/portability-data-transfer-microsoft/src/test/java/org/dataportabilityproject/transfer/microsoft/integration/LocalExportTestRunner.java:
```diff
@@ -0,0 +1,37 @@
+import org.dataportabilityproject.types.transfer.models.contacts.ContactsModelWrapper;
+
+/**
+ *
```
fill in comment

holachuy commented on extensions/data-transfer/portability-data-transfer-microsoft/src/test/java/org/dataportabilityproject/transfer/microsoft/integration/LocalExportTestRunner.java:
```diff
@@ -0,0 +1,37 @@
+package org.dataportabilityproject.transfer.microsoft.integration;
```
license header

holachuy commented on extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/types/GraphPagination.java:
```diff
@@ -0,0 +1,25 @@
+ */
+@JsonTypeName("org.dataportability:GraphPagination")
+public class GraphPagination extends PaginationData {
+    private String nextLink;
```
final

holachuy commented on extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/types/GraphPagination.java:
```diff
@@ -0,0 +1,25 @@
+package org.dataportabilityproject.transfer.microsoft.types;
```
license header

holachuy commented on extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/transformer/contacts/ToVCardTransformer.java:
```diff
@@ -0,0 +1,134 @@
+
+
+/*
+Investigate:
```
TODO

holachuy commented on extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/transformer/contacts/ToVCardTransformer.java:
```diff
@@ -0,0 +1,134 @@
+package org.dataportabilityproject.transfer.microsoft.transformer.contacts;
```
license header

holachuy commented on extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/transformer/contacts/ToVCardAddressTransformer.java:
```diff
@@ -0,0 +1,25 @@
+package org.dataportabilityproject.transfer.microsoft.transformer.contacts;
```
license header

rtannenbaum commented on extensions/cloud/portability-cloud-microsoft/src/main/java/org/dataportabilityproject/cloud/microsoft/cosmos/CosmosStore.java:
```diff
@@ -0,0 +1,154 @@
+    }
+
+    @Override
+    public <T extends DataModel> T findData(Class<T> type, String id) {
```
As I'm porting the Google cloud solution over to JobStore, I'm implementing "PortabilityJob find(String jobId)" directly and don't seem to need the generic findData.

Since we're only storing PortabilityJobs in the store as of now, can we get rid of the generic findData?

rtannenbaum replied:
same thing goes for the other generic type methods - create and update. Can we just have the PortabilityJob versions?

jimmarino replied:
I was planning that as a generic way for importers/exporters to store data but maybe the mechanism should just be ContinuationData? What do you think about having JobStore persist continuation data, or is there a better way to handle that?

On a related note, there is a streaming version of the persistence method on JobStore as well. My thought there was to provide a mechanism to directly stream large amounts of data from the exporter to the store and from the store to the importer based on a key that can be provided by continuation data. 

Chuy and I has a discussion about this a few weeks back and we talked about the fact that we need some form of streaming storage since the exporters and importers may be running in difference processes/containers and hence the input and output streams cannot be directly chained.

jimmarino replied:
I should also add that if the JobStore persists all job-related data, it could be that multiple backing services are used. For example in the Microsoft implementation, I was planning on using Cosmos DB for the Job data and the separate Blob Storage service for large data. The Microsoft implementation of JobStore would composite those two services but clients won't need to know about that.

rtannenbaum replied:
ah ok, I haven't been following the ContinuationData discussion so I'm not sure, but would it be OK to leave as a PortabilityJob store at first so it's easier to migrate the existing impls over, then we can add ContinuationData / streaming later as necessary?

This is what I had in mind: https://github.com/google/data-portability/pull/112

jimmarino replied:
Sounds good to me!

rtannenbaum commented on extensions/auth/portability-auth-microsoft/src/main/java/org/dataportabilityproject/auth/microsoft/package-info.java:
```diff
@@ -16,4 +16,5 @@
 /**
  * Implements importers and exporters for Google services.
```
nit: still says Google in the documentation

jimmarino commented:
OK I rebased my commits against master and fixed all of the items highlighted in the pull request comments. Let me know if there are any outstanding issues. 

Thanks for reviewing and the suggestions.

holachuy commented:
HI Jim,

Thanks, I hope you had a great weekend!

I don't believe I'm seeing the latest in GitHub, I only see my latest round
of comments from last week (license headers, final modifiers, and other
suggestions)

Thx

On Mon, Feb 5, 2018 at 4:21 AM, Jim Marino <notifications@github.com> wrote:

> OK I rebased my commits against master and fixed all of the items
> highlighted in the pull request comments. Let me know if there are any
> outstanding issues.
>
> Thanks for reviewing and the suggestions.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/data-portability/pull/104#issuecomment-363041672>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AYzhd1zZVJ90upoDYID86uax2a9BZ6tuks5tRtYpgaJpZM4RyksQ>
> .
>



-- 


Jessie 'Chuy' Chavez | Software Engineer | Chicago Tech50
<http://www.chicagobusiness.com/article/20150715/NEWS08/150719954/meet-jessie-chuy-chavez-a-google-engineer-who-leads-by-example>
 | chuy@google.com | 773-458-6683

jimmarino commented:
Hi Chuy,

I rebased my commits in the branch so the commit history will remain clean. The changes are pushed to my branch - do you see them in there?
